docs/Distributed.md

```diff
@@ -0,0 +1,212 @@
+   For example, on Linux this is accessible with: `openssl rand -hex 64`.
+ - When relying on random number generators (for generating keys or
+   tokens), be aware that a lack of entropy is possible in cloud or
+   virtualized environments in some scenarios.
+
+== Overview
+
+Distributed sccache consists of three parts:
+
+ - the client, an sccache binary that wishes to perform a compilation on
+   remote machines
+ - the scheduler (`sccache-dist` binary), responsible for deciding where
+   a compilation job should run
+ - the server (`sccache-dist` binary), responsible for actually executing
+   a build
+
+All servers are required to be a 64-bit Linux install. Clients may request
+compilation from Linux, Windows or OSX. Linux compilations will attempt to
+automatically package the compiler in use, while Windows and OSX users will
+need to specify a toolchain for cross-compilation ahead of time.
+
+== Communication
+
+The HTTP implementation of sccache has the following API:
+
+ - scheduler
+   - `POST alloc_job`
+      - Called by a client to submit a compilation request.
+      - Returns information on where the job is allocated it should run.
+   - `POST heartbeat_server`
+      - Called (repeatedly) by servers to register as available for jobs.
+   - `POST state`
+      - Called by servers to inform the scheduler of the state of the job.
+   - `GET status`
+      - Returns information about the scheduler.
+ - `server`
+   - `POST assign_job`
+      - Called by the scheduler to inform of a new job being assigned to this server.
+      - Returns whether the toolchain is already on the server or needs submitting.
+   - `POST submit_toolchain`
+      - Called by the client to submit a toolchain.
+   - `POST run_job`
+      - Called by the client to run a job.
+      - Returns the compilation stdout along with files created.
+
+There are two axes of security in this setup:
+
+1. Can the scheduler trust the servers?
+2. Is the client permitted to submit and run jobs?
+
+=== Server Trust
+
+If a server is malicious, they can return malicious compilation output to a user.
+To protect against this, servers must be authenticated. You have three means for
+doing this, and the scheduler and all servers must use the same mechanism.
+
+Once a server has registered itself using the selected authentication, the scheduler
+will trust the registered server address and use it for builds.
+
+==== JWT HS256 (preferred)
+
+This method uses secret key to create a per-IP-and-port token for each server.
+Acquiring a token will only allow participation as a server if the attacker can
+additionally impersonate the IP and port the token was generated for.
+
+You *must* keep the secret key safe.
+
+*To use it*:
+
+Create a scheduler key with `sccache-dist auth generate-jwt-hs256-key` (which will
+use your OS random number generator) and put it in your scheduler config file as
+follows:
+
+```
+server_auth = { type = "jwt_hs256", secret_key = "YOUR_KEY_HERE" }
```

Comment:
It would be great to have examples of full config files for each of the scheduler, client, builder in the docs at some point.